core/src/main/java/org/apache/iceberg/V1Metadata.java

```diff
@@ -356,7 +356,7 @@ public Object get(int pos) {
 
     @Override
     public void put(int i, Object v) {
-      throw new UnsupportedOperationException("Cannot read into IndexedDataFile");
+      throw new UnsupportedOperationException("Cannot write into IndexedDataFile");
```

Comment:
This doesn't add NaN counts to v1 metadata, but there's no reason why we can't. Adding a new column to v1 metadata files is forward compatible and is allowed, so we can go ahead and add it here. That way, even v1 benefits.

The reason why this is a v2 feature is that we are creating additional requirements for writers: NaN must not be written as a lower or upper bound in Iceberg metadata. Writers that want to write NaN counts in v1 should be able to do so.

Reply:
Sure! I wasn't sure if I wanted to add it v1 since v1 was an established version, and I don't know if introducing new attributes would be considered as breaking people's assumption on the file contract (e.g. if some people have their own version of manifest entry parser). But I guess the chance of this happening would probably be rare, and should be easy to fix.